pkg/shipper/shipper_test.go

```diff
@@ -0,0 +1,116 @@
+package shipper
+
+import (
+	"testing"
+
+	"context"
+	"io/ioutil"
+	"time"
+
+	"fmt"
+	"os"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/promlts/pkg/testutil"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/tsdb/labels"
+)
+
+type inMemStorage struct {
+	t    *testing.T
+	dirs map[string]struct{}
+}
+
+func newInMemStorage(t *testing.T) *inMemStorage {
+	return &inMemStorage{
+		t:    t,
+		dirs: make(map[string]struct{}),
+	}
+}
+
+func (r *inMemStorage) Exists(_ context.Context, id string) (bool, error) {
+	_, exists := r.dirs[id]
+	return exists, nil
+}
+
+func (r *inMemStorage) Upload(_ context.Context, dir string) error {
+	// Double check if shipper checks Exists method properly.
+	_, exists := r.dirs[dir]
+	testutil.Assert(r.t, !exists, "target should not exists")
+
+	r.dirs[dir] = struct{}{}
+	return nil
+}
+
+func TestShipper_UploadsBlocksFromProm(t *testing.T) {
+	logger := log.NewNopLogger()
+
+	dir, err := ioutil.TempDir("", "shipper-test-snapshots")
+	testutil.Ok(t, err)
+	defer os.RemoveAll(dir)
+
+	p, err := testutil.NewTSDB()
+	testutil.Ok(t, err)
+	defer p.Close()
+
+	storage := newInMemStorage(t)
+	shipper := New(
+		logger,
+		nil,
+		dir,
+		storage,
+		IsULIDDir,
+	)
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+
+	// Start shipping blocks.
+	go func() { testutil.Ok(t, shipper.Run(ctx)) }()
+
+	baseT := timestamp.FromTime(time.Now()) / 1000 * 1000
+
+	// Produce 10 different blocks and wait max 5 seconds for shipper to ship them.
+	for i := int64(0); i < 10; i++ {
+		ctx, cancel2 := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel2()
+
+		// Add fake intput.
+		a := p.Appender()
+
+		// Make labels, timestamp and values unique.
+		labelVal := fmt.Sprintf("b%v", i)
+		baseT := baseT + (i * 300)
+		baseV := float64(i * 3)
+
+		a.Add(labels.FromStrings("a", labelVal), baseT+100, baseV+1)
+		a.Add(labels.FromStrings("a", labelVal), baseT+200, baseV+2)
+		a.Add(labels.FromStrings("a", labelVal), baseT+300, baseV+3)
+		testutil.Ok(t, a.Commit())
+
+		testutil.Ok(t, p.Snapshot(dir))
+		waitForShipment(t, ctx, storage, len(p.Blocks()))
```

Comment:
So `Snapshot()` creates a new sub-directory for each call to snapshot. The snapshotted blocks will also be overlapping in general. The shipper watches the normal data dir though and checks for top-level items that are ULID dirs.

So this should not work I think.

Reply:
Ah, you are snapshotting directly back into the main dir. Was confusing that with where we push snapshots in Prometheus.
I thought though that we even added a safety guard forbidding that call with the dir itself.

Reply:
Mhh https://github.com/prometheus/tsdb/blob/master/db.go#L584

Reply:
It is forbidden, I have special dir just for this. (: